### PR TITLE
Raise error for missing bounds in optimizer SNOBFIT

### LINKED ISSUE #8580: Bug using SNOBFIT in combination with UCCSD (VQE) / misleading function signature of optimizer SNOBFIT
When running a VQE algorithm with ansatz UCCSD the following error message shows up:
```
[...]
 File "C:\Users\poc\Anaconda3\envs\qiskit\lib\site-packages\qiskit_nature\algorithms\pes_samplers\bopes_sampler.py", line 175, in sample
    self._raw_results = self._run_points(points)
  File "C:\Users\poc\Anaconda3\envs\qiskit\lib\site-packages\qiskit_nature\algorithms\pes_samplers\bopes_sampler.py", line 201, in _run_points
    raw_result = self._run_single_point(point)  # dict of results
  File "C:\Users\poc\Anaconda3\envs\qiskit\lib\site-packages\qiskit_nature\algorithms\pes_samplers\bopes_sampler.py", line 259, in _run_single_point
    result = self._state_solver.solve(self._problem, aux_ops_current_step)
  File "C:\Users\poc\Anaconda3\envs\qiskit\lib\site-packages\qiskit_nature\algorithms\ground_state_solvers\ground_state_eigensolver.py", line 91, in solve
    raw_mes_result = self._solver.compute_minimum_eigenvalue(main_operator, aux_ops)  # type: ignore
  File "C:\Users\poc\Anaconda3\envs\qiskit\lib\site-packages\qiskit\algorithms\minimum_eigen_solvers\vqe.py", line 530, in compute_minimum_eigenvalue
    opt_result = self.optimizer.minimize(
  File "C:\Users\poc\Anaconda3\envs\qiskit\lib\site-packages\qiskit\algorithms\optimizers\snobfit.py", line 96, in minimize
    if abs(theta) > bounds[idx][0]:
TypeError: '>' not supported between instances of 'float' and 'NoneType'
```

In case of ansatz `EfficientSU2` the code runs fine, only with `UCCSD` I get the error.

I tracked down the error to be caused by missing parameter bounds. The function [signature](https://github.com/Qiskit/qiskit-terra/blob/cbb64ed266b55525738de20a4e43419d5fef6e32/qiskit/algorithms/optimizers/snobfit.py#L82) of the optimizer's `minimize` function (``) suggests that `None` is a valid value. But the code never handles None-type input (neither `None` nor the nested list/tuple with `None` values).

So, I propose one of the following changes:
- Change the signature of the `minimize` method (still not easy to read for a user)
- Catch the `None`s in the input bounds and add a more verbose error message that optimizer SNOBFIT requires parameter bounds
- Artificially add bounds for the optimizer. Basically the parameter can only be varied from 0 to 2pi so one could catch the `None` input and change it to these bounds. Though I'm not aware of the side effects this approach might have... Here it might be useful to ask also someone from qiskit nature whether it could be resolved within the UCCSD ansatz (though it would not solve the initial problem with the misleading signature/missing error message).

Any fixes (or suggestions therefore) are highly appreciated.

### PR DESCRIPTION
**Pull request by cpossel**
### Summary

Raise error for invalid `None`s in bounds (i.e. missing bounds) of optimizer SNOBFIT.
Closes #8580 

### Details and comments

Function signature is misleading, allowing None for bounds. The resulting error is non-intuitive. So, this PR adds raising a more concise error message. This includes also `None` values within the bounds list (can happen e.g. with ansatz UCCSD).

### CODE REVIEW

**Comment by Cryoris** [comment]
Thanks for the contribution @cpossel! Could you add a test that checks this error is correctly raised? You could add it to
https://github.com/Qiskit/qiskit-terra/blob/main/test/python/algorithms/optimizers/test_optimizers.py

and check e.g. the following for an example of how to check that a code raises an error:
https://github.com/Qiskit/qiskit-terra/blob/2eee56616d50a9e26756f855ef4aa0135920ad78/test/python/algorithms/test_entangler_map.py#L51

**Comment by cpossel** [comment]
@Cryoris Test for failing SNOBFIT is added.
I didn't add any specific succeeding SNOBFIT test as it should already be captured (more or less) by the general scipy optimizer tests in the test suite.

**Comment by cpossel** [comment]
I need to add the scikit-quant package to the requirements file in order to fix the error `qiskit.exceptions.MissingOptionalLibraryError: "The 'scikit-quant' library is required to use 'SNOBFIT'. You can install it with 'pip install scikit-quant'."` in the terra testing environment. Should it be added to the [requirements-dev.txt](https://github.com/Qiskit/qiskit-terra/blob/main/requirements-dev.txt) file? Or is there another location to add this requirement to?

**Comment by woodsp-ibm** [comment]
There is a unit test for the scikit-quant optimizers BOBYQA, SNOBFIT, IMFIL https://github.com/Qiskit/qiskit-terra/blob/main/test/python/algorithms/optimizers/test_optimizers_scikitquant.py but I don't think we have ever run it in CI. Any additional test should be in this file.

The documentation notes this on scikit-quant:

https://github.com/Qiskit/qiskit-terra/blob/a0db5817ac39222abeaef647a38df491d688d221/qiskit/algorithms/optimizers/__init__.py#L87-L92

**Comment by cpossel** [comment]
The test cases in test_optimizers.py follow the logic to test the proper functioning of the optimizer itself while test_optimizers_scikitquant.py uses a holistic approach i.e. testing for the correct output of a VQE calculation with the chosen optimizer.
So, following this logic, the SNOBFIT test I want to add fits far better into test_optimizers.py. I'd also like to point out that the are already 2 scipy optimizer tests (unsing BFGS)  in test_optimizers.py, namely `test_scipy_optimizer` and `test_scipy_optimizer_callback`.

test_optimizers_scikitquant.py uses the `RealAmplitudes` ansatz which sets the bounds by itself, so the initial error can't be reproduced with the test template in test_optimizers_scikitquant.py and multiple lines of code from test_optimizers.py would need to be copied to test_optimizers_scikitquant.py what seems very redundant and unintuitive for me.

Making the test case run without installing the third party package would still be possible catching the `MissingOptionalLibraryError` error the same way as done in test_optimizers_scikitquant.py

**Comment by woodsp-ibm** [comment]
I am not sure I am fully following. The unit test method is just checking the bounds. All you need is to call minimize on it and pass some dummy objective function since it should fail immediately right. The run_optimizer is way more than is needed since that was done to check the optimizer could find the minimum of a function to check its correct working. Your test just checks that it fails as expected given an invalid bounds on the minimize right. To me it would be better to include that in the test suite for those 3. Yes it does use VQE - maybe finding some more straightforward function did not test the optimizers adequately. Ideally we would use as little other code as was needed to check their correctness - but I know sometimes things just don't work out.

File: test/python/algorithms/optimizers/test_optimizers.py [hunk 1]
```diff
@@ -38,6 +38,7 @@
     P_BFGS,
     POWELL,
     SLSQP,
+    SNOBFIT,
```
Comment (woodsp-ibm) [inline test/python/algorithms/optimizers/test_optimizers.py]: This line importing SNOBFIT should go too, to put things back as they were

Comment (cpossel) [reply test/python/algorithms/optimizers/test_optimizers.py]: I saw it the moment I pushed the commit :D

**Comment by cpossel** [comment]
As far as I understand the output of the CI tests the test's failure is due to some CI configuration and not due to the code changes. 
So, can anyone of you @Cryoris @woodsp-ibm @jakelishman look into the issue of the failed test, please?

**Comment by jakelishman** [comment]
The tests are all passing now - I see that the last commit had a test failure that's since been re-run, and it was just caused by a dodgy network connection / overallocated hardware on Azure's side.  Nothing to worry about - it's just transient and we don't have control over it.

**Comment by woodsp-ibm** [comment]
Looks good, but can you add a [release note](https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md#release-notes) for this bug fix.

**Review by woodsp-ibm — APPROVED** [review]
Thanks for the contribution. All looks good to me now.

**Pull request merged**